File: fp_py/src/main/fruitpunch/native/physics/physics_world.py
from fruitpunch.native.elements.element import Element
import libfp_core
from fruitpunch.native.native_class import NativeClass

class PhysicsWorld(Element):
    '''
    A container for processing physics and detect collision events.     
    '''

    def __init__(self):
        NativeClass.__init__(self)
        
        self.store(libfp_core.PhysicsWorld())        

    # ---------------------------------------------------------------------
    # Native Methods
    # ---------------------------------------------------------------------

    # _____________________________________________________________________

    def init(self):
        '''
        Makes the physics world ready to accept bodies and trigger 
        contact events.
        '''

        return self.run_native("init")
    # _____________________________________________________________________

    def render_frame(self):
        '''
        Updates the physical world.
        Executed every frame, and is responsible for executing a single 
        step that updates the status of the bodies, contacts, and joints in 
        the physics world. The framerate can be set by setting the timeStep 
        member variable.
        '''

        return self.run_native("render_frame")

    def get_bodies(self):
        return self.run_native("get_bodies")
    # _____________________________________________________________________

    def get_gravity(self):
        return self.run_native("get_gravity")
    # _____________________________________________________________________

    def get_joints(self):
        return self.run_native("get_joints")
    # _____________________________________________________________________

    def get_pixel_size(self):
        return self.run_native("get_pixel_size")
    # _____________________________________________________________________

    def get_position_iterations(self):
        return self.run_native("get_position_iterations")
    # _____________________________________________________________________

    def get_shapers(self):
        return self.run_native("get_shapers")
    # _____________________________________________________________________

    def get_timestep(self):
        return self.run_native("get_timestep")
    # _____________________________________________________________________

    def get_velocity_iterations(self):
        return self.run_native("get_velocity_iterations")
    # _____________________________________________________________________

    def get_world(self):
        return self.run_native("get_world")
    # _____________________________________________________________________

    def get_bodies_sensor_by_default(self):
        return self.run_native("get_bodies_sensor_by_default")
    # _____________________________________________________________________
    
    def set_bodies(self, bodies):
        return self.run_native("set_bodies", bodies)
    # _____________________________________________________________________

    def set_gravity(self, gravity):
        p = libfp_core.Point()
        p.x = gravity.x
        p.y = gravity.y
        
        return self.run_native("set_gravity", p)
    # _____________________________________________________________________

    def set_pixel_size(self, pixelSize):
        return self.run_native("set_pixel_size", pixelSize)
    # _____________________________________________________________________

    def set_position_iterations(self, positionIterations):
        return self.run_native("set_position_iterations", positionIterations)
    # _____________________________________________________________________

    def set_timestep(self, timestep):
        return self.run_native("set_timestep", timestep)
    # _____________________________________________________________________

    def set_velocity_iterations(self, velocityIterations):
        return self.run_native("set_velocity_iterations", velocityIterations)
    
    # ______________________________________________________________________
    
    def set_bodies_sensor_by_default(self, bodiesSensorByDefault):
        print("set_bodies_sensor_by_default called")
        return self.run_native("set_bodies_sensor_by_default", bodiesSensorByDefault)

    # ---------------------------------------------------------------------
    # Properties
    # ---------------------------------------------------------------------

    bodies = property(get_bodies, set_bodies)
    gravity = property(get_gravity, set_gravity)
    joints = property(get_joints)
    pixel_size = property(get_pixel_size, set_pixel_size)
    position_iterations = property(get_position_iterations, set_position_iterations)
    shapers = property(get_shapers)
    timestep = property(get_timestep, set_timestep)
    velocity_iterations = property(get_velocity_iterations, set_velocity_iterations)
    bodies_sensor_by_default = property(get_bodies_sensor_by_default, set_bodies_sensor_by_default)
    world = property(get_world)

NativeClass.class_map[libfp_core.PhysicsWorld] = PhysicsWorld

File: fp_core/src/main/src/Physics/PhysicsWorld.cpp
/*
 * PhysicsWorld.cpp
 *
 *  Created on: Sep 9, 2012
 *      Author: leo
 */

#include <fruitpunch/Physics/PhysicsWorld.h>
#include <fruitpunch/Physics/Shaper.h>
#include <fruitpunch/Physics/Body.h>
#include <fruitpunch/Physics/SpriteShaper.h>
#include <fruitpunch/Physics/ContactEvent.h>
#include <fruitpunch/Common/Events.h>
#include <Box2D/Box2D.h>
#include <boost/shared_ptr.hpp>
#include <cstdio>

using namespace std;

namespace fp_core {
namespace physics {

boost::shared_ptr<PhysicsWorld> PhysicsWorld::mCurrent;

PhysicsWorld::PhysicsWorld() : Element() {
	point p;
	p.x = 0;
	p.y = 10.0;
	mGravity = p;
	mTimestep = 1.0f/60.0f;
	mPositionIterations = 6;
	mVelocityIterations = 3;
	mPixelSize = 15.0f;
	mBodiesSensorByDefault = false;

}

// ____________________________________________________________________________

PhysicsWorld::~PhysicsWorld() {

}

// ____________________________________________________________________________

void PhysicsWorld::init() {
	b2Vec2 gravity(mGravity.x, mGravity.y);
	mWorld = boost::shared_ptr<b2World>(new b2World(gravity));
	mWorld->SetContactListener(this);
	mCurrent = getThis<PhysicsWorld>();

	mShapers.push_back(SpriteShaper::ref(new SpriteShaper()));

}

// ____________________________________________________________________________

void PhysicsWorld::renderFrame() {
	mWorld->Step(mTimestep, mVelocityIterations, mPositionIterations);
}

// ____________________________________________________________________________

boost::shared_ptr<PhysicsShape> PhysicsWorld::makeShapeFor(
		Element::ref element) {
	vector<Shaper::ref>::iterator it = mShapers.begin();
	for(; it != mShapers.end(); it++) {
		if ((*it)->test(element)) {
			return (*it)->make(element,
					getThis<PhysicsWorld>());
		}
	}
	return boost::shared_ptr<PhysicsShape>();
}

// ____________________________________________________________________________

void PhysicsWorld::triggerContact(b2Contact& contact, std::string eventName) {
	ContactEvent::ref event(new ContactEvent(contact, eventName, getThis<PhysicsWorld>()));
	this->notify(event);
}

// ____________________________________________________________________________

void PhysicsWorld::onAdd() {
	init();
}

// ____________________________________________________________________________

void PhysicsWorld::BeginContact(b2Contact* contact) {
	triggerContact(*contact, EV_BEGIN_CONTACT);
}

// ____________________________________________________________________________

void PhysicsWorld::EndContact(b2Contact* contact) {
	triggerContact(*contact, EV_END_CONTACT);
}

// ____________________________________________________________________________

void PhysicsWorld::PreSolve(b2Contact* contact, const b2Manifold* oldManifold) {
	triggerContact(*contact, EV_PRE_SOLVE);
}

// ____________________________________________________________________________

void PhysicsWorld::PostSolve(b2Contact* contact,
		const b2ContactImpulse* impulse) {
	triggerContact(*contact, EV_POST_SOLVE);
}

// ____________________________________________________________________________

void PhysicsWorld::addFixtureToElementMapping(
		boost::shared_ptr<Element> element, b2Fixture* fixture) {
	mFixtureElementMapping.insert(pair<b2Fixture *, boost::weak_ptr<Element> >(fixture, element));
}

// ____________________________________________________________________________

void PhysicsWorld::removeFixtureToElementMapping(b2Fixture* fixture) {
	map<b2Fixture *, boost::weak_ptr<Element> >::iterator it;
	it = mFixtureElementMapping.find(fixture);

	mFixtureElementMapping.erase(it);
}

// ____________________________________________________________________________

boost::shared_ptr<Element> PhysicsWorld::getElementFromFixture(
		b2Fixture* fixture) {

	map<b2Fixture *, boost::weak_ptr<Element> >::iterator it;
	it = mFixtureElementMapping.find(fixture);

	// could not find a fixture. Return null.
	if (it == mFixtureElementMapping.end()) {
		return boost::shared_ptr<Element>();
	}

	return it->second.lock();
}

// ____________________________________________________________________________

b2Fixture* PhysicsWorld::getFixtureFromElement(
		boost::shared_ptr<Element> element) {
	map<b2Fixture *, boost::weak_ptr<Element> >::iterator it = mFixtureElementMapping.begin();
	for (; it != mFixtureElementMapping.end(); it++) {
		if (it->second.lock() == element) {
			return it->first;
		}
	}

	return NULL;
}


// ____________________________________________________________________________

const boost::shared_ptr<std::vector<Body> > PhysicsWorld::getBodies() const {
	return mBodies;
}

void PhysicsWorld::setBodies(
		const boost::shared_ptr<std::vector<Body> > bodies) {
	mBodies = bodies;
}

boost::shared_ptr<PhysicsWorld> PhysicsWorld::getCurrent() {
	return mCurrent;
}

void PhysicsWorld::setCurrent(boost::shared_ptr<PhysicsWorld> current) {
	mCurrent = current;
}

point PhysicsWorld::getGravity() const {
	return mGravity;
}

void PhysicsWorld::setGravity(point gravity) {
	mGravity = gravity;
	b2Vec2 grav;
	grav.x = gravity.x;
	grav.y = gravity.y;
	mWorld->SetGravity(grav);
}

const std::vector<boost::shared_ptr<PhysicsJoint> > PhysicsWorld::getJoints() const {
	return mJoints;
}

float PhysicsWorld::getPixelSize() const {
	return mPixelSize;
}

void PhysicsWorld::setPixelSize(float pixelSize) {
	mPixelSize = pixelSize;
}

int PhysicsWorld::getPositionIterations() const {
	return mPositionIterations;
}

void PhysicsWorld::setPositionIterations(int positionIterations) {
	mPositionIterations = positionIterations;
}

const std::vector<boost::shared_ptr<Shaper> >& PhysicsWorld::getShapers() const {
	return mShapers;
}

float PhysicsWorld::getTimestep() const {
	return mTimestep;
}

void PhysicsWorld::setTimestep(float timestep) {
	mTimestep = timestep;
}

int PhysicsWorld::getVelocityIterations() const {
	return mVelocityIterations;
}

void PhysicsWorld::setVelocityIterations(int velocityIterations) {
	mVelocityIterations = velocityIterations;
}

const boost::shared_ptr<b2World> PhysicsWorld::getWorld() const {
	return mWorld;
}

bool PhysicsWorld::getBodiesSensorByDefault() const {
	return mBodiesSensorByDefault;
}

void PhysicsWorld::setBodiesSensorByDefault(bool isSensorDefault) {
	mBodiesSensorByDefault = isSensorDefault;
}

} /* namespace physiscs */
} /* namespace fp_core */
